[diem-vm] Clean up unused mut, refactored out the api for executing single transaction

runtian-zhou opened the pull request:
…ingle transaction.

<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Diem project.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

Currently there's a number of unused mutable references in the diem_vm. This PR cleans up the usused `mut` and also refactored out the api for executing single transaction. This would be useful for our future experiements of transaction parallel execution.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/master/CONTRIBUTING.md#pull-requests)?
Yes

## Test Plan

Should be a clean refactor. Need to pass all existing CIs.

bob-wilson commented on language/diem-vm/src/diem_transaction_executor.rs:
```diff
@@ -753,7 +764,7 @@ fn is_reconfiguration(vm_output: &TransactionOutput) -> bool {
 /// but a user transaction or writeset transaction is transformed to a SignatureCheckedTransaction.
 #[derive(Debug)]
-enum PreprocessedTransaction {
+pub(crate) enum PreprocessedTransaction {
```
What's this change for? I'm guessing so you can experiment with calling execute_single_transaction from elsewhere in the crate, and I guess that's OK temporarily. I'm not a big fan of this enum, though -- it's unnecessarily confusing the way that it renames BlockMetadata to BlockPrologue and GenesisTransaction to WaypointWriteSet. I'd like to see us get rid of this instead, but as long as we keep it within the crate, I don't object.

runtian-zhou replied:
That is exactly right. My intention for this PR is to separate out a clean api for executing one single transaction so that I can implement my parallel execution refactor on top.

runtian-zhou commented:
Renamed the variant for BlockMetadata per @bob-wilson 's request. I also changed the definition of `PreprocessedTransaction` to include a `InvalidSignature` variant to get rid of the nasty `Result<PreprocessedTransaction, VMStatus>` in the api.

bob-wilson commented:
> Renamed the variant for BlockMetadata per @bob-wilson 's request. I also changed the definition of `PreprocessedTransaction` to include a `InvalidSignature` variant to get rid of the nasty `Result<PreprocessedTransaction, VMStatus>` in the api.

I was thinking to get rid of the PreprocessedTransaction enum entirely, but OK. In general, I think it would be better to make changes like that in separate PRs. Sorry if it sounded like I was asking you to do something about it now.....

runtian-zhou commented:
/land
